tls/s2n_tls13_certificate_verify.c

```diff
@@ -0,0 +1,194 @@
+    DEFER_CLEANUP(struct s2n_hash_state message_hash = {0}, s2n_hash_free);
+    DEFER_CLEANUP(struct s2n_stuffer unsigned_content = {0}, s2n_stuffer_free);
+    GUARD(s2n_hash_new(&message_hash));
+    GUARD(s2n_hash_init(&message_hash, chosen_sig_scheme->hash_alg));
+
+    uint32_t maximum_signature_length = s2n_pkey_size(pkey);
+    GUARD(s2n_alloc(&signed_content, maximum_signature_length));
+    signed_content.size = maximum_signature_length;
+
+    if (conn->mode == S2N_CLIENT) {
+        GUARD(s2n_tls13_generate_unsigned_cert_verify_content(conn, &unsigned_content, S2N_CLIENT));
+    } else {
+        GUARD(s2n_tls13_generate_unsigned_cert_verify_content(conn, &unsigned_content, S2N_SERVER));
+    }
+
+    GUARD(s2n_hash_update(&message_hash, unsigned_content.blob.data, s2n_stuffer_data_available(&unsigned_content)));
+    GUARD(s2n_pkey_sign(pkey, chosen_sig_scheme->sig_alg, &message_hash, &signed_content));
+
+    GUARD(s2n_stuffer_write_uint16(out, signed_content.size));
+    GUARD(s2n_stuffer_write_bytes(out, signed_content.data, signed_content.size));
+
+    return 0;
+}
+
+int s2n_tls13_generate_unsigned_cert_verify_content(struct s2n_connection *conn, struct s2n_stuffer *unsigned_content, s2n_mode mode)
+{
+    s2n_tls13_connection_keys(tls13_ctx, conn);
+
+    struct s2n_hash_state handshake_hash, hash_copy;
+    uint8_t hash_digest_length = tls13_ctx.size;
+    uint8_t digest_out[S2N_MAX_DIGEST_LEN];
+
+    /* Get current handshake hash */
+    GUARD(s2n_handshake_get_hash_state(conn, tls13_ctx.hash_algorithm, &handshake_hash));
+
+    /* Copy current hash content */
+    GUARD(s2n_hash_new(&hash_copy));
+    GUARD(s2n_hash_copy(&hash_copy, &handshake_hash));
+    GUARD(s2n_hash_digest(&hash_copy, digest_out, hash_digest_length));
+    GUARD(s2n_hash_free(&hash_copy));
+
+    /* Concatenate the content to be signed/verified */
+    GUARD(s2n_stuffer_alloc(unsigned_content, hash_digest_length + s2n_tls13_cert_verify_header_length(mode)));
+    GUARD(s2n_stuffer_write_bytes(unsigned_content, S2N_CERT_VERIFY_PREFIX, sizeof(S2N_CERT_VERIFY_PREFIX)));
+
+    if (mode == S2N_CLIENT) {
+        GUARD(s2n_stuffer_write_bytes(unsigned_content, S2N_CLIENT_CERT_VERIFY_CONTEXT, sizeof(S2N_CLIENT_CERT_VERIFY_CONTEXT)));
+    } else {
+        GUARD(s2n_stuffer_write_bytes(unsigned_content, S2N_SERVER_CERT_VERIFY_CONTEXT, sizeof(S2N_SERVER_CERT_VERIFY_CONTEXT)));
+    }
+
+    GUARD(s2n_stuffer_write_bytes(unsigned_content, digest_out, hash_digest_length));
+
+    return 0;
+}
+
+uint8_t s2n_tls13_cert_verify_header_length(s2n_mode mode)
+{
+    if (mode == S2N_CLIENT) {
+        return sizeof(S2N_CERT_VERIFY_PREFIX) + sizeof(S2N_CLIENT_CERT_VERIFY_CONTEXT);
+    }
+    return sizeof(S2N_CERT_VERIFY_PREFIX) + sizeof(S2N_SERVER_CERT_VERIFY_CONTEXT);
+}
+
+int s2n_tls13_cert_verify_recv(struct s2n_connection *conn)
+{
+    struct s2n_signature_scheme chosen_sig_scheme = s2n_rsa_pkcs1_md5_sha1;
```

Comment:
Although this would work, I still think this should point to `conn->secure.conn_sig_scheme` or `conn->secure.client_cert_sig_scheme` depending on the conn's mode as this mirrors s2n_tls13_cert_verify_send().

eg. when a server sends a cert verify to the client, the client's `secure.conn_sig_scheme` scheme never gets correctly until receiving this message, likewise a server's `client_cert_sig_scheme` is never set correctly until the server receives a client cert verify.